Cython/Utility/ObjectHandling.c

```diff
@@ -465,7 +465,9 @@ static CYTHON_INLINE PyObject *__Pyx_GetItemInt_Fast(PyObject *o, Py_ssize_t i,
         }
     }
 #else
-    if (is_list || PySequence_Check(o)) {
+    // PySequence_GetItem behaves differently to PyObject_GetItem for i<0
+    // and possibly some other cases so can't generally be substituted
+    if (is_list || PyList_CheckExact(o) || PyTuple_CheckExact(o)) {
```

Comment:
Here I've just made a fast path for `list` and `tuple`. I suspect it could be extended to include a few other builtin classes (`bytes`, `unicode`, `bytearray`, `memoryview`) but I didn't do that to keep the check simple.

Reply:
Maybe like this?
```suggestion
    if (is_list || !PyMapping_Check(o)) {
```
The intention is to avoid the `int` object creation. If we assume (!) that checking for the protocol is faster than that, we win for sequences. But we'd pay with a double check for mappings. CPU-L1 caching might still keep that efficient in the end.

Note that this is all guessing. There are multiple settings in which this code will execute, PyPy, Limited API, user configurations in CPython, … I really cannot make a good guess for all of them, and guessing about performance is bound to go wrong anyway.